src/adapter/src/coord/sequencer/inner/subscribe.rs

```diff
@@ -0,0 +1,411 @@
+                    return;
+                }
+                Timestamp(stage) => {
+                    let next = return_if_err!(self.subscribe_timestamp(&mut ctx, stage).await, ctx);
+                    (ctx, SubscribeStage::OptimizeLir(next))
+                }
+                OptimizeLir(stage) => {
+                    self.subscribe_optimize_lir(ctx, stage, otel_ctx);
+                    return;
+                }
+                Finish(stage) => {
+                    let result = self.subscribe_finish(&mut ctx, stage).await;
+                    ctx.retire(result);
+                    return;
+                }
+            }
+        }
+    }
+
+    fn subscribe_validate(
+        &mut self,
+        session: &mut Session,
+        SubscribeValidate {
+            plan,
+            target_cluster,
+        }: SubscribeValidate,
+    ) -> Result<SubscribeOptimizeMir, AdapterError> {
+        let plan::SubscribePlan { from, when, .. } = &plan;
+
+        let cluster = self
+            .catalog()
+            .resolve_target_cluster(target_cluster, session)?;
+        let cluster_id = cluster.id;
+
+        let mut replica_id = session
+            .vars()
+            .cluster_replica()
+            .map(|name| {
+                cluster
+                    .replica_id(name)
+                    .ok_or(AdapterError::UnknownClusterReplica {
+                        cluster_name: cluster.name.clone(),
+                        replica_name: name.to_string(),
+                    })
+            })
+            .transpose()?;
+
+        // SUBSCRIBE AS OF, similar to peeks, doesn't need to worry about transaction
+        // timestamp semantics.
+        if when == &QueryWhen::Immediately {
+            // If this isn't a SUBSCRIBE AS OF, the SUBSCRIBE can be in a transaction if it's the
+            // only operation.
+            session.add_transaction_ops(TransactionOps::Subscribe)?;
+        }
+
+        let depends_on = from.depends_on();
+
+        // Run `check_log_reads` and emit notices.
+        let notices = check_log_reads(
+            self.catalog(),
+            cluster,
+            &depends_on,
+            &mut replica_id,
+            session.vars(),
+        )?;
+        session.add_notices(notices);
+
+        // Determine timeline.
+        let mut timeline = self.validate_timeline_context(depends_on.clone())?;
+        if matches!(timeline, TimelineContext::TimestampIndependent) && from.contains_temporal() {
+            // If the from IDs are timestamp independent but the query contains temporal functions
+            // then the timeline context needs to be upgraded to timestamp dependent.
+            timeline = TimelineContext::TimestampDependent;
+        }
+
+        // TODO: this check is present in `sequence_create_index` / `sequence_create_materialized_view`
+        // Should we have it here as well?
+        // self.ensure_cluster_can_host_compute_item(name, *cluster_id)?;
```

Comment:
@jkosh44 I think you reviewed the PR that added this (#21846). LMK if it makes sense to add this check for `SUBSCRIBE` statements as well.

Reply:
Good call out, but is this somehow related to moving optimization off of the main Coord thread? If not I'd prefer to move it to a separate PR.